Modules/blake2module.c

```diff
@@ -0,0 +1,924 @@
+    data: object(c_default="NULL") = b''
+    /
+    *
+    digest_size: int(c_default="HACL_HASH_BLAKE2S_OUT_BYTES") = _blake2.blake2s.MAX_DIGEST_SIZE
+    key: Py_buffer(c_default="NULL", py_default="b''") = None
+    salt: Py_buffer(c_default="NULL", py_default="b''") = None
+    person: Py_buffer(c_default="NULL", py_default="b''") = None
+    fanout: int = 1
+    depth: int = 1
+    leaf_size: unsigned_long = 0
+    node_offset: unsigned_long_long = 0
+    node_depth: int = 0
+    inner_size: int = 0
+    last_node: bool = False
+    usedforsecurity: bool = True
+
+Return a new BLAKE2s hash object.
+[clinic start generated code]*/
+
+static PyObject *
+py_blake2s_new_impl(PyTypeObject *type, PyObject *data, int digest_size,
+                    Py_buffer *key, Py_buffer *salt, Py_buffer *person,
+                    int fanout, int depth, unsigned long leaf_size,
+                    unsigned long long node_offset, int node_depth,
+                    int inner_size, int last_node, int usedforsecurity)
+/*[clinic end generated code: output=556181f73905c686 input=8165a11980eac7f3]*/
+{
+    return py_blake2b_or_s_new(type, data, digest_size, key, salt, person, fanout, depth, leaf_size, node_offset, node_depth, inner_size, last_node, usedforsecurity);
+}
+
+/*[clinic input]
+_blake2.blake2b.copy
+
+Return a copy of the hash object.
+[clinic start generated code]*/
+
+static PyObject *
+_blake2_blake2b_copy_impl(Blake2Object *self)
+/*[clinic end generated code: output=622d1c56b91c50d8 input=e383c2d199fd8a2e]*/
+{
+    Blake2Object *cpy;
+
+    if ((cpy = new_Blake2Object(Py_TYPE(self))) == NULL)
+        return NULL;
+
+    ENTER_HASHLIB(self);
+    switch (self->impl) {
+#if HACL_CAN_COMPILE_SIMD256
+        case Blake2b_256:
+            cpy->blake2b_256_state = Hacl_Hash_Blake2b_Simd256_copy(self->blake2b_256_state);
+            break;
+#endif
+#if HACL_CAN_COMPILE_SIMD128
+        case Blake2s_128:
+            cpy->blake2s_128_state = Hacl_Hash_Blake2s_Simd128_copy(self->blake2s_128_state);
+            break;
+#endif
+        case Blake2b:
+            cpy->blake2b_state = Hacl_Hash_Blake2b_copy(self->blake2b_state);
+            break;
+        case Blake2s:
+            cpy->blake2s_state = Hacl_Hash_Blake2s_copy(self->blake2s_state);
+            break;
+        default:
+            Py_UNREACHABLE();
+    }
+    cpy->impl = self->impl;
+    LEAVE_HASHLIB(self);
+    return (PyObject *)cpy;
+}
+
+/*[clinic input]
+_blake2.blake2b.update
+
+    data: object
+    /
+
+Update this hash object's state with the provided bytes-like object.
+[clinic start generated code]*/
+
+static PyObject *
+_blake2_blake2b_update(Blake2Object *self, PyObject *data)
+/*[clinic end generated code: output=e6d1ac88471df308 input=ffc4aa6a6a225d31]*/
+{
+    Py_buffer buf;
+
+    GET_BUFFER_VIEW_OR_ERROUT(data, &buf);
+
+    if (!self->use_mutex && buf.len >= HASHLIB_GIL_MINSIZE) {
+        self->use_mutex = true;
+    }
```

Comment:
Once you've set the mutex, does it mean that you will always use one? because, maybe the next time, you'll use a smaller data.

Reply:
@gpshead would know the answer to this one -- this is the pattern used in sha2module.c so I didn't introduce anything new here

Reply:
being consistent with how we do it in other hash functions is appropriate, that was a good one to copy from.  IIRC the reasoning behind this upgrade-to-mutex-only state is that realistically people calling update on a large enough chunk of data are going to continue to do so, perhaps until only their final call to update which _might_ be smaller.  there's no point in trying to the mutex for that one final call in an update-feeding loop.  The logic required to _safely_ return to not using one again would be complicated concurrency wise for no real benefit.